estimagic/differentiation/numdiff_np.py

```diff
@@ -0,0 +1,240 @@
+            to a rule of thumb as long as this does not conflict with min_steps.
+        scaling_factor (np.ndarray or float): Scaling factor which is applied to
+            base_steps. If it is an np.ndarray, it needs to have the same shape as x.
+            scaling_factor is useful if you want to increase or decrease the base_step
+            relative to the rule-of-thumb or user provided base_step, for example to
+            benchmark the effect of the step size. Default 1.
+        lower_bounds (np.ndarray): 1d array with lower bounds for each parameter.
+        upper_bounds (np.ndarray): 1d array with upper bounds for each parameter.
+        step_ratio (float or array): Ratio between two consecutive Richardson
+            extrapolation steps in the same direction. default 2.0. Has to be larger
+            than one. step ratio is only used if n_steps > 1.
+        min_steps (np.ndarray): Minimal possible step sizes that can be chosen to
+            accommodate bounds. Needs to have same length as x. By default min_steps is
+            equal to base_steps, i.e step size is not decreased beyond what is optimal
+            according to the rule of thumb.
+        f0 (np.ndarray): 1d numpy array with func(x), optional.
+        n_cores (int): Number of processes used to parallelize the function
+            evaluations. Default 1.
+
+    Returns:
+        derivative (np.ndarray): The estimated first derivative of func at x.
+            The shape of the output depends on the dimension of x and func(x):
+            f: R -> R leads to shape (1,), usually called derivative
+            f: R^m -> R leads to shape (m, ), usually called Gradient
+            f: R -> R^n leads to shape (n, 1), usually called Jacobian
+            f: R^m -> R^n leads to shape (n, m), usually called Jacobian
+
+    """
+    func_kwargs = {} if func_kwargs is None else func_kwargs
+    partialed_func = functools.partial(func, **func_kwargs)
+    f0 = partialed_func(x) if f0 is None else f0
+
+    x_was_scalar = np.isscalar(x)
+    f_was_scalar = np.isscalar(f0)
+
+    x = np.atleast_1d(x)
+    f0 = np.atleast_1d(f0)
+
+    @nan_if_exception
+    @de_scalarize(x_was_scalar)
+    def internal_func(x):
+        return partialed_func(x)
+
+    steps = generate_steps(
+        x=x,
+        method=method,
+        n_steps=n_steps,
+        target="first_derivative",
+        base_steps=base_steps,
+        scaling_factor=scaling_factor,
+        lower_bounds=lower_bounds,
+        upper_bounds=upper_bounds,
+        step_ratio=step_ratio,
+        min_steps=min_steps,
+    )
+
+    evaluation_points = []
+    for step_arr in steps:
+        for i, j in product(range(n_steps), range(len(x))):
+            if np.isfinite(step_arr[i, j]):
+                point = x.copy()
+                point[j] += step_arr[i, j]
+                evaluation_points.append(point)
+            else:
+                evaluation_points.append(np.nan)
```

Comment:
Explicit NaN check and then changed the order instead `not np.isnan`.

```suggestion
            if np.isnan(step_arr[i, j]):
                evaluation_points.append(np.nan)
            else:
                point = x.copy()
                point[j] += step_arr[i, j]
                evaluation_points.append(point)
```